dynamic_programming/fibonacci_sequence_recursion_with_memoization.py

```diff
@@ -0,0 +1,49 @@
+# Fibonacci Sequence Using Recursion and Memoization
+
+# Memoization Concept: https://en.wikipedia.org/wiki/Memoization
+# Fibonacci Concept : https://pt.wikipedia.org/wiki/Sequ%C3%AAncia_de_Fibonacci
+
+
+def fibonacci(n: int, cache: dict = dict()) -> int:
```

Comment:
Cache is an implementation detail and not something which needs to be exposed to the client. So, we should not use that as a function parameter and instead make a variable `cache` either globally or if using a class, a private instance variable.

Another thing is you need to provide type hints for the dictionary keys and values like so: `dict[keytype, valuetype]`

Also, never keep a mutable object as the default argument value. It can have unforeseen side effects.

Reply:
Corrected, thanks for the remarks